Improve support for `--entryPointStrategy Packages`

akphi opened the pull request:
This PR provides some improvements for users of `--entryPointStrategy Packages`. These includes:

- [x] Improved support for `--entryPointStrategy Packages`. TypeDoc will now load package-specific configurations from `package.json` `typedoc` field. This configuration allows configuring a custom display name (`typedoc.displayName`) field, entry point (`typedoc.entryPoint` - this is equivalent and will override `typedocMain`), and path to a readme file to be rendered at the top of the package page (`typedoc.readmeFile`), fixes #1658 
-  [x] The `--includeVersion` option will now be respected by `--entryPointStrategy Packages`. Also, for this combination, missing `version` field in the root `package.json` will not issue a warning - _this fix would be useful for folks who want to turn on `--includeVersion` and yet don't really want to put a version in the main `package.json` because it's not always meaningful to do so in a monorepo_

This is the shape of the `typedoc` config (in the child-packages' `package.json`):

```ts
export interface TypedocPackageManifestConfig {
    entryPoint?: string | undefined;
    displayName?: string | undefined;
    readmeFile?: string | undefined;
}
```

I feel that having per-package config allows more flexibility.

This is a screenshot of my change (the README per module and the version per child package)


https://user-images.githubusercontent.com/13574879/177013017-8ddb2f9c-736a-4bb8-8bc3-a6cefee64d55.mov


Let me know what you think about these features!


---
**UPDATED**

I notice that when I develop locally, converter tests will likely fail due to the hardcoded `url` in the `specs.json` file. I suppose you don't mind since you run locally and push directly to the repo, for us who work on the fork, these tests will likely fail.

^ Thanks @Gerrit0 ! Works like charms now 🙏

Gerrit0 commented:
> due to the hardcoded url in the specs.json 

Shoot, didn't think about that when adding url to the serialized JSON... that's something I should fix.

akphi commented:
Also, I'm not so sure, if this is accepted, what's the best way to document this on the website. Is it gonne be through https://typedoc.org/guides/options/#entrypointstrategy ?

Gerrit0 commented:
This is big enough I need to wait until I'm on a computer to really take a look at it. I'm not entirely sure I like adding another property to declaration reflections, but it might be a good temporary workaround until 0.24 when adding support for arbitrary markdown documents.... should be able to look at it Wednesday at the latest

akphi commented:
> This is big enough I need to wait until I'm on a computer to really take a look at it. I'm not entirely sure I like adding another property to declaration reflections, but it might be a good temporary workaround until 0.24 when adding support for arbitrary markdown documents.... should be able to look at it Wednesday at the latest

Yep, that was my worry as well, since I copied it from `ProjectReflection`, but I don't think we have a separate thing like `PackageReflection` yet. Anyway, glad I'm not totally off 😄

Gerrit0 commented on src/lib/converter/plugins/PackagePlugin.ts:
```diff
@@ -140,9 +143,15 @@ export class PackagePlugin extends ConverterComponent {
+                    // since not all monorepo specifies a meaningful version to the main package.json
+                    // this warning should be optional
+                    if (
+                        this.entryPointStrategy !== EntryPointStrategy.Packages
```
Should we also add this check in the `includeVersion` check below this block?

akphi replied:
Yes, that's a good catch! Updated

Gerrit0 commented on src/lib/output/themes/default/partials/index.tsx:
```diff
@@ -67,8 +67,19 @@ export function index(context: DefaultThemeRenderContext, props: ContainerReflec
+            {props instanceof DeclarationReflection &&
+                props.kind === ReflectionKind.Module &&
+                props.readme?.length !== 0 && (
+                    <section class="tsd-panel-group tsd-index-group">
```
This should definitely not be a `tsd-index-group`, that class refers to the page index, it isn't just a generic container.

akphi replied:
I have removed `tsd-index-group`

Gerrit0 commented on src/lib/utils/entry-point.ts:
```diff
@@ -384,7 +391,23 @@ function getEntryPointsForPackages(
+            displayName:
+                typedocPackageConfig?.displayName ??
+                // if displayName is not configured, use the package name (and version, if configured)
+                `${packageJson["name"]}${
```
I have a few concerns here:

1. Including the version in reflection names means that cross-module links must specify the version in the link... resulting in broken links when bumping a version (`foo@1.2.3!bar` vs `foo!bar`)
2. `displayName` gives another way to rename exports, which is already possible through [`@module`](https://typedoc.org/tags/module/)... is there an advantage to this?
3. We should probably warn if trying to add a version and there's no version.
4. The formatting is inconsistent with how `PackagePlugin` sets the name if there is a version. I'd like these to be the same.
5. Even if `displayName` is set, if `includeVersion` is specified, we should probably add the version.

akphi replied:
1. I have not thought about this case. I will fix it so that we don't end up changing the reflection name, but just the rendering name - I will try.
2. I didn't know about this and I just tried and I found a limitation with this. That is the `@module` comment must be the first block comment above any code. Unfortunately, we can't make use of this since we have our copyright header. See this example: https://github.com/finos/legend-studio/blob/master/packages/legend-art/src/index.ts
3. Will do
4. Will do
5. Will do

akphi replied:
@Gerrit0 I think to carry around information on `version` I will need to poke into reflection model once again, I'll add a field for `version` there. Are you okay with this? Or is there some other way that I'm not familiar with. Also, I'm trying to do a fair bit in this one PR. My main concern is to allow specifying the `README` per child packages. So if this change for `displayName` and `version` pose to be more complicated, I think I can leave them out of this PR altogether.

Gerrit0 replied:
2 - Makes sense! It stays.
version - Yep, again I don't love it, but it's a reasonable solution.

Gerrit0 commented on src/lib/utils/entry-point.ts:
```diff
@@ -384,7 +391,23 @@ function getEntryPointsForPackages(
+                        ? `@${packageJson["version"]}`
+                        : ""
+                }`,
+            readmeFile: typedocPackageConfig?.readmeFile
```
Any reason not to default this to `README.md` in the same directory as the `package.json`?... in fact, do you have any examples in the wild where this would *not* be the (case insensitive) desired file to read?

akphi replied:
Yeah, I base this off the `readme` field in `typedoc` config. I think the default there is `none`. Also, when I think of this feature, one use case that makes sense to me is that we can have a directory somewhere with a bunch of markdown files, and the `readmeFile` link per child package can point of those files instead of the `README.md` file in each pacakge directory. That way, I can collate documents to some folder rather than having to scatter them in each package, for example:


```
documentation
  - doc1.md
  - doc2.md
packages
  - pkg1
    - package.json: typedoc.readmeFile --> ../../documentation/doc1.md
  - pkg2
```

Well, the one thing that I'm concerned about is the name of the config `readmeFile`, which seems to mislead people into thinking that this is pointing at the `README`, the way I think about this is more like a custom doc page per package.

Gerrit0 replied:
> I think the default there is none

The default is to discover a readme file starting at the common directory of all source files and working up the file system.

> the way I think about this is more like a custom doc page per package.

The reason I'm willing to let this in today is *because* it's a README file for a package and is intentionally limited in scope. The big feature for 0.24 is planned to be arbitrary extra markdown files in a project, but this is definitely not that. `readmeFile` is a good name IMO.

akphi replied:
so I guess I don't need to do anything here then? I'm going ahead to resolve this.

Gerrit0 commented on src/lib/utils/package-manifest.ts:
```diff
@@ -36,6 +36,54 @@ export function loadPackageManifest(
+    ) {
+        // TODO: we should consider another less manual approach in the future,
+        // e.g. uses a library to deserialize config from raw JSON
+        const config: TypedocPackageManifestConfig = packageJson.typedoc;
```
What about `readmeFile`?

akphi replied:
I missed it, will add that now

Gerrit0 commented on src/lib/utils/package-manifest.ts:
```diff
@@ -36,6 +36,54 @@ export function loadPackageManifest(
+        typeof packageJson.typedoc == "object" &&
+        packageJson.typedoc
+    ) {
+        // TODO: we should consider another less manual approach in the future,
```
There is `src/lib/utils/validation.ts`, this file was written before it existed and hasn't been updated to use it where it makes sense, see `src/test/utils/validation.ts` for usage examples.

akphi replied:
Great!! That's a really nice approach :). Updated!

Gerrit0 replied:
I think it makes sense to be strict today - there's no way to specify options here, so no plugins would expect to put stuff in this object.

Gerrit0 commented on src/lib/models/reflections/declaration.ts:
```diff
@@ -129,6 +130,11 @@ export class DeclarationReflection extends ContainerReflection {
+    /**
+     * The contents of the readme.md file of the module when found.
+     */
+    readme?: CommentDisplayPart[];
```
I still don't love this, but am willing to live with it until 0.24.

akphi commented on src/lib/utils/package-manifest.ts:
```diff
@@ -205,7 +253,13 @@ export function getTsEntryPointForPackage(
+    if (typedocPackageConfig?.entryPoint) {
+        packageMain = typedocPackageConfig.entryPoint;
+    } else if (
         hasOwnProperty(packageJson, "typedocMain") &&
```
I will add the warning. I don't think there would be case where `typedocMain` is preferred, because I'm for making `typedoc` config a structure in `package.json`.

akphi replied:
Added tests for this as well

akphi commented:
@Gerrit0 I have addressed most of the PR comments, except for the ones related to `displayName` and `version` in `entry-point.ts` file, I want to hear your opinion about that approach so I could move forward.

akphi commented:
To clear out the reviews comment while I work on them, I have taken the liberty to resolve some of them, I hope you're okay with this. I will `unresolve` them once I finish fixing the code 🙏

Gerrit0 commented on src/lib/utils/package-manifest.ts:
```diff
@@ -205,10 +247,19 @@ export function getTsEntryPointForPackage(
         typeof packageJson.typedocMain == "string"
     ) {
+        logger.warn(
+            `Legacy typedoc entry point config (using "typedocMain" field) found for "${packageJsonPath}". Please update to use "typedoc": { "entryPoint": "..." } instead. In future upgrade, "typedocMain" field will be ignored.`
```
I'd like to pass the path through `nicePath`, which is exported from utils/fs, which will shorten paths if possible.

akphi replied:
Thanks!

Gerrit0 commented:
>  I have taken the liberty to resolve some of them

Go right ahead :) I always do a full pass before approving anyways, so this won't result in missing anything.

akphi commented:
I believe we have resolved all the review comments. Thank you so much @Gerrit0 !

Gerrit0 commented:
Thank you! Almost all of the monorepo support in TypeDoc has come from someone who uses them adding support, so it's great to get PRs from people that use them
